File: cpp_solutions/LinkedList/Remove_Duplicates_from_Sorted_list_II.cpp
/*
* Given a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list.
*
* For example,
* Given 1->2->3->3->4->4->5, return 1->2->5.
* Given 1->1->1->2->3, return 2->3.
*
**/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode *root = NULL;
        ListNode **ppNext = &root;
        while(head) {
            if(head->next == NULL || head->next->val != head->val) {
                *ppNext = head;
                ppNext = &(head->next);
                head = head->next;
            } else {
                int val = head->val;
                do {
                    head = head->next;
                } while(head != NULL && head->val == val);
            }
        }
        *ppNext = NULL;
        return root;
    }
};

File: python_solutions/LinkedList/Remove_Duplicates_from_Sorted_List_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head is None or head.next is None:
            return head
        # pHead, pPrev -> [1,2,3,3,4,4,5]
        pPrev, pNode = ListNode(-1), head
        pPrev.next, pHead = pNode, pPrev
        while pNode and pNode.next:
            if pNode.val == pNode.next.val:
                val = pNode.val
                while pNode and val == pNode.val:
                    pNode = pNode.next
                pPrev.next = pNode
            else:
                pNode = pNode.next
                pPrev = pPrev.next
        return pHead.next